### PR TITLE
[SYCL][ESIMD] Add support for transposed prefetch for 1/2 byte elements

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -2851,24 +2878,54 @@ prefetch_impl(AccessorTy acc, OffsetT byte_offset, simd_mask<1> pred) {
                 "Implicit truncation of 64-bit byte_offset to 32-bit is "
                 "disabled. Use -fsycl-esimd-force-stateless-mem or explicitly "
                 "convert offsets to a 32-bit vector");
-  check_lsc_vector_size<NElts>();
   check_lsc_data_size<T, DS>();
   check_cache_hints<cache_action::prefetch, PropertyListT>();
+
+  constexpr size_t Alignment =
+      detail::getPropertyValue<PropertyListT, alignment_key>(sizeof(T));
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:2885]: This line does not seem correct. For bytes/words the alignment should still be 4-bytes or more.
for block_load the caller of block_load_impl adds the default alignment 4-or-more to the property list: https://github.com/intel/llvm/blob/602fddfb8f9f64db00e0072f233a0eef1be4ae27/sycl/include/sycl/ext/intel/esimd/memory.hpp#L1353-L1355

I think both functions prefetch_impl and LSC-block_load_impl should expect that alignment is already added to the properties. Can you please do this for prefetch_impl and block_load_impl:
```suggestion
  constexpr size_t Alignment = get_property<alignment_key>().value;
```

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp [hunk 2]
```diff
@@ -378,3 +396,164 @@ template <typename T> bool testACC(queue Q) {
...
+  // Only d/q-words are supported now.
+  // Thus we use this I32Factor for testing purposes and convenience.
+  constexpr int I32Factor =
+      std::max(static_cast<int>(sizeof(int) / sizeof(T)), 1);
+  Passed &= testBlockLoadPrefetchUSM<T, 1 * I32Factor>(Q, 2, 4, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 2 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 4 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 8 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 16 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 32 * I32Factor>(Q, 2, 4, DG2OrPVCProps);
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp:476]: Please check that 256-elements for words and 256/512-elements for bytes can be prefetched.
That may require splitting of the prefetch_usm_dg2_pvc.cpp to 2 tests - 1 for DG2 and 1 for PVC.

**Review by v-klochkov — COMMENTED** [review]
Thank you for enabling byte/word element prefetch. It looks good in general. I have only few comments.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -8876,15 +8933,24 @@ prefetch(const T *p, OffsetSimdViewT byte_offsets, PropertyListT props = {}) {
 /// Prefetches elements of the type 'T' from continuous memory location
 /// addressed by the base pointer \p p, and offset \p byte_offset and the length
 /// \p VS elements into the cache.
+/// The maximum size of prefetched block is 512 bytes for PVC and 256 bytes for
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8936]:
```suggestion
/// The maximum size of a prefetched block is 512 bytes for PVC and 256 bytes for
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 4]
```diff
@@ -8876,15 +8933,24 @@ prefetch(const T *p, OffsetSimdViewT byte_offsets, PropertyListT props = {}) {
...
+/// ACM (DG2). When sizeof(T) equal to 8 the address must be 8-byte aligned.
+/// Also, 8-bytes alignment is required when the function has to load more than
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:8938]:
```suggestion
/// Also, 8-byte alignment is required when the function has to load more than
```

**Review by v-klochkov — APPROVED** [review]

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:2889]: Sounds reasonable. Thanks for adding the extra assert to prefetch_impl().

File: sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp [hunk 5]
```diff
@@ -459,21 +460,28 @@ template <typename T> bool testBlockLoadPrefetchUSM(queue Q) {
...
-  Passed &= testBlockLoadPrefetchUSM<T, 8 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
-  Passed &= testBlockLoadPrefetchUSM<T, 16 * I32Factor>(Q, 5, 5, DG2OrPVCProps);
-  Passed &= testBlockLoadPrefetchUSM<T, 32 * I32Factor>(Q, 2, 4, DG2OrPVCProps);
+  Passed &= testBlockLoadPrefetchUSM<T, 1 * I32Factor>(Q, 2, 4, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 2 * I32Factor>(Q, 5, 5, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 4 * I32Factor>(Q, 5, 5, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 8 * I32Factor>(Q, 5, 5, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 16 * I32Factor>(Q, 5, 5, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 32 * I32Factor>(Q, 2, 4, Props);
+  Passed &= testBlockLoadPrefetchUSM<T, 64 * I32Factor>(Q, 2, 4, Props);
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp:481]: The `int64 x 64` case must be called for PVC only

Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/unified_memory_api/Inputs/prefetch.hpp:481]: It causes  JIT build failure on DG2 CI bot.

**Pull request merged**